[package] gmp/6.3.0: build error on cppstd 23 due to change in function declaration grammar

### Description

C++ 23 changed the function declaration syntax such that:
void g(){}
no longer means, function with unknown parameters and now reads as function that takes no parameters. Therefore in C++ 23:
void g(){}
is no equivalent to
void g(void){}

Current GMP relies on this in a single test, causing the build to fail if compiled from source on C++ 23, as seen in the [mailing list](https://gmplib.org/list-archives/gmp-bugs/2024-November/005550.html)

### Package and Environment Details

* Package Name/Version: **gmp/6.3.0**
* Operating System+version: **Arch Linux**
* Compiler+version: **Clang 19.1.7**
* Conan version: **conan 2.16.1**
* Python version: **Python 3.13.3**


### Conan profile

[settings]
arch=x86_64
compiler=clang
compiler.cppstd=gnu23
compiler.libcxx=libc++
compiler.version=19
os=Linux
[buildenv]
C=clang
CXX=clang++


### Steps to reproduce

conan install --requires=gmp/6.3.0 --build=missing

### Logs

Build output
<details><summary>Click to expand log</summary>

```
-------- Installing package gmp/6.3.0 (2 of 2) --------
gmp/6.3.0: Building from source
gmp/6.3.0: Package gmp/6.3.0:2743db357664f5d00d503290edc626b45f57311c
gmp/6.3.0: settings: os=Linux arch=x86_64 compiler=gcc compiler.cppstd=gnu17 compiler.libcxx=libstdc++11 compiler.version=14 build_type=Release
gmp/6.3.0: options: disable_assembly=True enable_cxx=True enable_fat=False fPIC=True shared=False
gmp/6.3.0: Copying sources to build folder
gmp/6.3.0: Building your package in /home/nicobuzeta/.conan2/p/b/gmp93c8e24a60bf1/b
gmp/6.3.0: Calling generate()
gmp/6.3.0: Generators folder: /home/nicobuzeta/.conan2/p/b/gmp93c8e24a60bf1/b/build-release/conan
gmp/6.3.0: Generating aggregated env files
gmp/6.3.0: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
gmp/6.3.0: Calling build()
gmp/6.3.0: Apply patch (file): patches/0001-msvc-dumpbin-yasm-wrapper.patch
gmp/6.3.0: Apply patch (file): patches/6.x.x-0001-fix-MSVC-next-prime-error.patch
gmp/6.3.0: RUN: "/home/nicobuzeta/.conan2/p/b/gmp93c8e24a60bf1/b/src/configure" --disable-shared --enable-static --prefix=/ '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' --with-pic=yes --enable-assembly=no --enable-fat=no --enable-cxx=yes --srcdir=../src
checking build system type... zen4-pc-linux-gnu
checking host system type... zen4-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking ABI=64
checking compiler gcc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
checking whether cc is gcc... yes
checking compiler cc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
checking ABI=x32
checking compiler gcc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
checking whether cc is gcc... yes
checking compiler cc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
checking ABI=32
checking compiler gcc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
checking compiler icc  -m64 -fPIC -O3  -DNDEBUG... no
checking whether cc is gcc... yes
checking compiler cc  -m64 -fPIC -O3  -DNDEBUG... no, long long reliability test 1
configure: error: could not find a working compiler, see config.log for details

gmp/6.3.0: ERROR:
Package '2743db357664f5d00d503290edc626b45f57311c' build failed
gmp/6.3.0: WARN: Build folder /home/nicobuzeta/.conan2/p/b/gmp93c8e24a60bf1/b/build-release
ERROR: gmp/6.3.0: Error in build() method, line 152
        autotools.configure()
        ConanException: Error 1 while executing

```

</details>

config.log

<details><summary>Click to expand log</summary>

```
conftest.c: In function 'f':
conftest.c:12:48: error: too many arguments to function 'g'; expected 0, have 6
   12 | for(i=0;i<1;i++){if(e(got,got,9,d[i].n)==0)h();g(i,d[i].src,d[i].n,got,d[i].want,9);if(d[i].n)h();}}
      |                                                ^ ~
conftest.c:7:6: note: declared here
    7 | void g(){}
      |      ^
```

</details>

--- a/recipes/gmp/all/conandata.yml
+++ b/recipes/gmp/all/conandata.yml
@@ -19,6 +19,7 @@ patches:
   "6.3.0":
     - patch_file: "patches/0001-msvc-dumpbin-yasm-wrapper.patch"
     - patch_file: "patches/6.x.x-0001-fix-MSVC-next-prime-error.patch"
+    - patch_file: "patches/cpp23-unprototyped-functions.patch"
     - patch_file: "patches/emscripten/0001-configure.patch"
       patch_os: "Emscripten"
       patch_source: "https://github.com/fpelliccioni"

--- a/recipes/gmp/all/patches/cpp23-unprototyped-functions.patch
+++ b/recipes/gmp/all/patches/cpp23-unprototyped-functions.patch
@@ -1,0 +1,12 @@
+--- configure
++++ configure
+@@ -6568,7 +6568,7 @@
+
+ #if defined (__GNUC__) && ! defined (__cplusplus)
+ typedef unsigned long long t1;typedef t1*t2;
+-void g(){}
++void g(int,t1 const*,t1,t2,t1 const*,int){}
+ void h(){}
+ static __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
+ {t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}
+

--- a/recipes/gmp/all/patches/cpp23-unprototyped-functions.patch
+++ b/recipes/gmp/all/patches/cpp23-unprototyped-functions.patch
@@ -5,7 +5,7 @@
  #if defined (__GNUC__) && ! defined (__cplusplus)
  typedef unsigned long long t1;typedef t1*t2;
 -void g(){}
-+void g(int,t1 const*,t1,t2,t1 const*,int){}
++void g(int a,t1 const* b,t1 c,t2 d,t1 const* e,int f){}
  void h(){}
  static __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
  {t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}

